src/Numerics/Mesh/Metrics.jl

```diff
@@ -98,127 +100,298 @@ function creategrid!(x1, x2, x3, e2c, ξ1, ξ2, ξ3)
 end
 
 """
-    computemetric!(x1, J, JcV, ξ1x1, sJ, n1, D)
+    setup_geom_factors_data!(vgeo, D)
+
+Input arguments:
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- D::DAT2, 1-D derivative operator on the device in the first dimension
+
+Setup the variable needed for geometric factors data in vgeo.
+"""
+function setup_geom_factors_data!(
+    vgeo::VolumeGeometry{NTuple{3,Int}, <:AbstractArray},
+    D
+)
+    Nq = size(D, 1)
+    nelem = div(length(vgeo.ωJ), Nq)
+
+    vgeo.x1 = reshape(vgeo.x1, (Nq, nelem))
+    vgeo.ωJ = reshape(vgeo.ωJ, (Nq, nelem))
+    vgeo.JcV = reshape(vgeo.JcV, (Nq, nelem))
+    vgeo.ξ1x1 = reshape(vgeo.ξ1x1, (Nq, nelem))
+
+end
+
+"""
+    setup_geom_factors_data!(vgeo, D1, D2)
+
+Input arguments:
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- D1::DAT2, 1-D derivative operator on the device in the first dimension
+- D2::DAT2, 1-D derivative operator on the device in the second dimension
+
+Setup the variable needed for geometric factors data in vgeo.
+"""
+function setup_geom_factors_data!(
+    vgeo::VolumeGeometry{NTuple{3,Int}, <:AbstractArray},
+    D1,
+    D2
+)
+    Nq = (size(D1, 1), size(D2, 1))
+    nelem = div(length(vgeo.ωJ), prod(Nq))
+
+    vgeo.x1 = reshape(vgeo.x1, (Nq..., nelem))
+    vgeo.x2 = reshape(vgeo.x2, (Nq..., nelem))
+    vgeo.ωJ = reshape(vgeo.ωJ, (Nq..., nelem))
+    vgeo.JcV = reshape(vgeo.JcV, (Nq..., nelem))
+    vgeo.ξ1x1 = reshape(vgeo.ξ1x1, (Nq..., nelem))
+    vgeo.ξ2x1 = reshape(vgeo.ξ2x1, (Nq..., nelem))
+    vgeo.ξ1x2 = reshape(vgeo.ξ1x2, (Nq..., nelem))
+    vgeo.ξ2x2 = reshape(vgeo.ξ2x2, (Nq..., nelem))
+
+end
+
+"""
+    setup_geom_factors_data!(vgeo, D1, D2, D3)
+
+Input arguments:
+- vgeo::VolumeGeometry, a struct containing the volumetric geometric factors
+- D1::DAT2, 1-D derivative operator on the device in the first dimension
+- D2::DAT2, 1-D derivative operator on the device in the second dimension
+- D3::DAT2, 1-D derivative operator on the device in the third dimension
+
+Setup the variable needed for geometric factors data in vgeo.
+"""
+function setup_geom_factors_data!(
+    vgeo::VolumeGeometry{NTuple{3,Int}, <:AbstractArray},
+    D1,
+    D2,
+    D3,
+)
+    Nq = (size(D1, 1), size(D2, 1), size(D3, 1))
+    nelem = div(length(vgeo.ωJ), prod(Nq))
+
+    vgeo.x1 = reshape(vgeo.x1, (Nq..., nelem))
```

Comment:
we can't modify the object a field of an immutable object refers to

Reply:
Thanks! I think it's taken care of now.